# Network Optimization: Case 9.1 (Aiding Allies)
[Source: Frederick S. Hillier and Gerald J. Lieberman.  *Introduction to Operations Research - 7th ed.*, McGraw-Hill, ISBN 0-07-232169-5]

Commander Votachev steps into the cold October night and deeply inhales the smoke from his cigarette, savoring its warmth. He surveys the destruction surrounding him — shattered windows, burning buildings, torn roads — and smiles. His two years of work training revolutionaries east of the Ural Mountains has proved successful; his troops now occupy seven strategically important cities in the Russian Federation: Kazan, Perm,Yekaterinburg, Ufa, Samara, Saratov, and Orenburg. His siege is not yet over, however. He looks to the west. Given the political and economic confusion in the Russian Federation at this time, he knows that his troops will be able to conquer Saint Petersburg and Moscow shortly. Commander Votachev will then be able to rule with the wisdom and control exhibited by his communist predecessors Lenin and Stalin.

Across the Pacific Ocean, a meeting of the top security and foreign policy advisers of the United States is in progress at the White House. The President has recently been briefed about the communist revolution masterminded by Commander Votachev and is determining a plan of action. The President reflects upon a similar October long ago in 1917, and he fears the possibility of a new age of radical Communist rule accompanied by chaos, bloodshed, escalating tensions, and possibly nuclear war. He therefore decides that the United States needs to respond and to respond quickly. Moscow has requested assistance from the United States military, and the President plans to send troops and supplies immediately.

The President turns to General Lankletter and asks him to describe the preparations being taken in the United States to send the necessary troops and supplies to the Russian Federation. General Lankletter informs the President that along with troops, weapons, ammunition, fuel, and supplies, aircraft, ships, and vehicles are being assembled at two port cities with airfields: Boston and Jacksonville. The aircraft and ships will transfer all troops and cargo across the Atlantic Ocean to the Eurasian continent. The general hands the President a list of the types of aircraft, ships, and vehicles being assembled along with a description of each type. The list is shown below.

|Transportation Type| Name| Capacity|Speed|
|:---|:---|:---|:---|
|Aircraft|C-141 Starlifter|150 tons|400 miles per hour|
|Ship|Transport|240 tons|35 miles per hour|
|Vehicle|Palletized Load System Truck|16,000 kilograms|60 miles per hour|

- All aircraft, ships, and vehicles are able to carry both troops and cargo.
- Once an aircraft or ship arrives in Europe, it stays there to support the armed forces.

The President then turns to Tabitha Neal, who has been negotiating with the NATO countries for the last several hours to use their ports and airfields as stops to refuel and resupply before heading to the Russian Federation. She informs the President that the following ports and airfields in the NATO countries will be made available to the United States military.

|Ports|Airfields|
|:---|:---|
|Napoli|London|
|Hamburg|Berlin|
|Rotterdam|Istanbul|


The President stands and walks to the map of the world projected on a large screen in the middle of the room. He maps the progress of troops and cargo from the United States to three strategic cities in the Russian Federation that have not yet been seized by Commander Votachev. The three cities are Saint Petersburg, Moscow, and Rostov. He explains that the troops and cargo will be used both to defend the Russian cities and to launch a counterattack against Votachev to recapture the cities he currently occupies. The President also explains that:

- All Starlifters and transports leave either Boston or Jacksonville.
- All transports that have traveled across the Atlantic must dock at one of the NATO ports to unload.
- Palletized load system trucks brought over in the transports will then carry all troops and materials unloaded from the ships at the NATO ports to the three strategic Russian cities not yet seized by Votachev - Saint Petersburg, Moscow, and Rostov.
- All Starlifters that have traveled across the Atlantic must land at one of the NATO airfields for refueling. The planes will then carry all troops and cargo from the NATO airfields to the three Russian cities.


<img src="images/aiding_allies_map.png" width="900" align="center">

The different routes that may be taken by the troops and supplies to reach the Russian Federation from the United States are shown below:

<img src="images/possible_routes_new.png" width="500" length="500" align="center">

## 1. The Shortest Path Problem: When time is of the essence
Moscow and Washington do not know when Commander Votachev will launch his next attack. Leaders from the two countries have therefore agreed that

- Troops should reach each of the three strategic Russian cities as quickly as possible.

The President has determined that the situation is so dire that cost is no object—as many Starlifters, transports, and trucks as are necessary will be used to transfer troops and cargo from the United States to Saint Petersburg, Moscow, and Rostov. Therefore,

- No limitations exist on the number of troops and amount of cargo that can be transferred between any cities.

The President has been given the following information about the length of the available routes between cities on the Atlantic leg:

|From|To|Length of route (km)|
|:---|:---|:---:|
|Boston |Berlin |7,250 |
|Boston |Hamburg |8,250 |
|Boston |Istanbul |8,300 |
|Boston |London |6,200 |
|Boston |Rotterdam |6,900 |
|Boston |Napoli |7,950 |
|Jacksonville |Berlin |9,200 |
|Jacksonville |Hamburg |9,800 |
|Jacksonville |Istanbul |10,100 |
|Jacksonville |London |7,900 |
|Jacksonville |Rotterdam |8,900 |
|Jacksonville |Napoli |9,400 |

and on the Eurasian leg:

|From|To|Length of route (km)|
|:---|:---|:---:|
|Berlin | Saint Petersburg |1,280 |
|Hamburg |Saint Petersburg |1,880 |
|Istanbul |Saint Petersburg |2,040 |
|London |Saint Petersburg |1,980 |
|Rotterdam |Saint Petersburg |2,200 |
|Napoli |Saint Petersburg |2,970 |
|Berlin |Moscow |1,600 |
|Hamburg |Moscow |2,120 |
|Istanbul |Moscow |1,700 |
|London |Moscow |2,300 |
|Rotterdam |Moscow |2,450 |
|Napoli |Moscow |2,890 |
|Berlin |Rostov |1,730 |
|Hamburg |Rostov |2,470 |
|Istanbul |Rostov |990 |
|London |Rostov |2,860 |
|Rotterdam |Rostov |2,760 |
|Napoli |Rostov |2,800 |

### Question

Given the distance and the speed of the transportation used between each pair of cities, how can the President most quickly move troops from the United States to each of the three strategic Russian cities? How long will it take troops and supplies to reach Saint Petersburg? Moscow? Rostov?

### 1.1 Dijkstra's shortest path approach
Since Question 1 is only concerned with the minimum time taken by the troops to reach Russia, we can model this problem as a *shortest path problem* with each edge length representing the time taken to travel from the source node to the target node. The problem can then be solved using *Dijkstra's shortest path algorithm*. We will use the helper classes and methods provided by the custom [solver](https://github.com/ayusbhar2/operations_research/tree/main/solver) module to solve the problem.

In [1]:
# my solver module
from solver.algorithms import get_shortest_path
from solver.classes import Edge, Graph
from solver.utils import get_result_summary, prettify

In [2]:
AIR_SPEED_KMPH = 400 * 1.60934  # from MPH to KMPH
WATER_SPEED_KMPH = 35 * 1.60934
LAND_SPEED_KMPH = 60 * 1.60934

In [3]:
# Create the network graph with provided edges
graph = Graph([
    Edge('Boston', 'Berlin', distance_km=7250, route_type='air'),
    Edge('Boston', 'Hamburg', distance_km=8250, route_type='sea'),
    Edge('Boston', 'Istanbul', distance_km=8300, route_type='air'),
    Edge('Boston', 'London', distance_km=6200, route_type='air'),
    Edge('Boston', 'Rotterdam', distance_km=6900, route_type='sea'),
    Edge('Boston', 'Napoli', distance_km=7950, route_type='sea'),
    Edge('Jacksonville', 'Berlin', distance_km=9200, route_type='air'),
    Edge('Jacksonville', 'Hamburg', distance_km=9800, route_type='sea'),
    Edge('Jacksonville', 'Istanbul', distance_km=10100, route_type='air'),
    Edge('Jacksonville', 'London', distance_km=7900, route_type='air'),
    Edge('Jacksonville', 'Rotterdam', distance_km=8900, route_type='sea'),
    Edge('Jacksonville', 'Napoli', distance_km=9400, route_type='sea'),
    
    Edge('Berlin', 'StPetersburg', distance_km=1280, route_type='land'),
    Edge('Hamburg', 'StPetersburg', distance_km=1880, route_type='land'),
    Edge('Istanbul', 'StPetersburg', distance_km=2040, route_type='air'),
    Edge('London', 'StPetersburg', distance_km=1980, route_type='air'),
    Edge('Rotterdam', 'StPetersburg', distance_km=2200, route_type='land'),
    Edge('Napoli', 'StPetersburg', distance_km=2970, route_type='land'),
    Edge('Berlin', 'Moscow', distance_km=1600, route_type='air'),
    Edge('Hamburg', 'Moscow', distance_km=2120, route_type='land'),
    Edge('Istanbul', 'Moscow', distance_km=1700, route_type='air'),
    Edge('London', 'Moscow', distance_km=2300, route_type='air'),
    Edge('Rotterdam', 'Moscow', distance_km=2450, route_type='land'),
    Edge('Napoli', 'Moscow', distance_km=2890, route_type='land'),
    Edge('Berlin', 'Rostov', distance_km=1730, route_type='air'),
    Edge('Hamburg', 'Rostov', distance_km=2470, route_type='land'),
    Edge('Istanbul', 'Rostov', distance_km=990, route_type='air'),
    Edge('London', 'Rostov', distance_km=2860, route_type='air'),
    Edge('Rotterdam', 'Rostov', distance_km=2760, route_type='land'),
    Edge('Napoli', 'Rostov', distance_km=2800, route_type='land'),
])

In [4]:
# Update all edges with a non-negative time-cost
for edge in graph.edges:
    if edge.route_type == 'air':
        time_hr = edge.distance_km / AIR_SPEED_KMPH
    elif edge.route_type == 'sea':
        time_hr = edge.distance_km / WATER_SPEED_KMPH
    else:
        time_hr = edge.distance_km / LAND_SPEED_KMPH
    edge.update(cost=time_hr) # edge cost is needed for shortest path algorithm

In [5]:
get_shortest_path(graph, 'Jacksonville', target='StPetersburg', algorithm='dijkstra')

(15.347906595250226, ['Jacksonville', 'London', 'StPetersburg'])

In [6]:
get_shortest_path(graph, 'Boston', target='StPetersburg', algorithm='dijkstra')

(12.707072464488547, ['Boston', 'London', 'StPetersburg'])

In [7]:
get_shortest_path(graph, 'Jacksonville', target='Moscow', algorithm='dijkstra')

(15.845004784570072, ['Jacksonville', 'London', 'Moscow'])

In [8]:
get_shortest_path(graph, 'Boston', target='Moscow', algorithm='dijkstra')

(13.204170653808394, ['Boston', 'London', 'Moscow'])

In [9]:
get_shortest_path(graph, 'Jacksonville', target='Rostov', algorithm='dijkstra')

(16.7149266158798, ['Jacksonville', 'London', 'Rostov'])

In [10]:
get_shortest_path(graph, 'Boston', target='Rostov', algorithm='dijkstra')

(13.949817937788161, ['Boston', 'Berlin', 'Rostov'])

From the above results we see that the fastest way to get troops and supplies to each of the three strategic Russian cities is:

- **Boston --> London --> Saint Petersburg (12.70 hr)**

- **Boston --> London --> Moscow (13.20 hr)**

- **Boston --> Berlin --> Rostov (13.95 hr)**

### 1.2 Binary Integer Programming approach

We can also model the above problem as a binary integer program as follows.

Parameters:

$$c_{j, k} := \text{ the "cost" associated with edge } (j, k)$$

Variables:

$$x_i:= \begin{cases}
1 \text{ if node i is on the path}\\
0 \text{ otherwise}
\end{cases}$$

$$y_{j, k}:= \begin{cases}
1 \text{ if edge (j, k) is on the path}\\
0 \text{ otherwise}
\end{cases}
$$

Objective:

$$\text{minimize} \sum_{j, k} y_{j, k} c_{j, k}$$

Constraints:

$$x_O := 1 \qquad\text{(Origin constraint)}$$

$$x_T := 1  \qquad\text{(Target constraint)}$$

$$
\sum_{j} y_{j, i} \le 1 + M (1 - x_i)\ , \quad i \ne O \quad \text{(Inbound constraint 1)}\\
\sum_{j} y_{j, i} \ge 1 - M (1 - x_i)\ , \quad i \ne O \quad \text{(Inbound constraint 2)}
$$

$$
\sum_{k} y_{i, k} \le 1 + M (1 - x_i)\ , \quad i \ne T \quad \text{(Outbound constraint 1)}\\
\sum_{k} y_{i, k} \ge 1 - M(1 - x_i)\ , \quad i \ne T \quad \text{(Outbound constraint 2)}
$$

$$
x_j + x_k \ge 2 y_{j, k} \quad \forall\ j, k \quad \text{(Connectivity constraint)}
$$

$$ x_i , y_{j, k} \quad \text{binary}
$$

In order to understand how this formulation would work, let us consider the subgraph with a single source node, viz. Boston, and a single target node, viz. St. Petersburg.

<img src="images/bip_formulation.png" width="500">

Lets solve the above program with `cvxpy`

In [11]:
import cvxpy as cp

In [12]:
def get_outbound_constraints(vertex, graph):
    s = '('
    t = '('
    for edge in graph.get_edges(source=vertex):
        s += 'y_{}_{} + '.format(edge.source, edge.target)
        t += 'y_{}_{} + '.format(edge.source, edge.target)
    s = s[:-2]
    s = s + ') <= 1 + M * (1 - x_{})'.format(vertex)
    
    t = t[:-2]
    t = t + ') >= 1 - M * (1 - x_{})'.format(vertex)
        
    return s, t

In [13]:
def get_inbound_constraints(vertex, graph):
    s = '('
    t = '('
    for edge in graph.get_edges(target=vertex):
        s += 'y_{}_{} + '.format(edge.source, edge.target)
        t += 'y_{}_{} + '.format(edge.source, edge.target)
    s = s[:-2]
    s = s + ') <= 1 + M * (1 - x_{})'.format(vertex)
    
    t = t[:-2]
    t = t + ') >= 1 - M * (1 - x_{})'.format(vertex)
        
    return s, t

In [14]:
def get_connectivity_constraints(edge):
    s = 'x_{s} + x_{t} >= 2 * y_{s}_{t}'.format(s=edge.source, t=edge.target)
    return s    

In [15]:
# Parameters

M = 1000000
for edge in graph.edges:
    exec('c_{s}_{t} = edge.cost'.format(s=edge.source, t=edge.target))

In [16]:
# Variables

variables = []
for edge in graph.edges:
    exec('y_{s}_{t} = cp.Variable(1, boolean=True, name="y_{s}_{t}")'.format(
        s=edge.source, t=edge.target))
    exec('variables.append(y_{s}_{t})'.format(s=edge.source, t=edge.target))

for vertex in graph.vertices:
    exec('x_{v} = cp.Variable(1, boolean=True, name="x_{v}")'.format(v=vertex))
    exec('variables.append(x_{v})'.format(v=vertex))

In [17]:
# Constraints

constraints_bip = [
    # origin and destination constraints
    x_Boston == 1,
    x_Jacksonville == 0,
    x_StPetersburg == 1,
    x_Moscow == 0,
    x_Rostov == 0,
]

# inbound and outbound constraints
for v in graph.vertices:
    if v != 'Boston' and v != 'Jacksonville':  # not a source node
        s, t = get_inbound_constraints(v, graph)
        constraints_bip.append(eval(s))
        constraints_bip.append(eval(t))
    if v != 'StPetersburg' and v != 'Moscow' and v != 'Rostov':  # not a target node
        s, t = get_outbound_constraints(v, graph)
        constraints_bip.append(eval(s))
        constraints_bip.append(eval(t))

# connectivity constraints
for edge in graph.edges:
    s = get_connectivity_constraints(edge)
    constraints_bip.append(eval(s))

In [18]:
# Objective

obj_str = ''
for edge in graph.edges:
    obj_str += 'c_{s}_{t} * y_{s}_{t} + '.format(s=edge.source, t=edge.target)

obj_str = obj_str[:-2]

objective_bip = cp.Minimize(eval(obj_str))

In [19]:
# Model

prob = cp.Problem(objective_bip, constraints_bip)

In [20]:
prob.solve();

In [21]:
summary = get_result_summary(prob)

In [23]:
summary['status']

'optimal'

In [24]:
summary['optimal_value']

12.707072464488547

In [26]:
df = prettify(summary['optimal_solution'])

In [29]:
df

variable  value
0                   x_Berlin    0.0
1                   x_Boston    1.0
2                  x_Hamburg    0.0
3                 x_Istanbul    0.0
4             x_Jacksonville    0.0
5                   x_London    1.0
6                   x_Moscow    0.0
7                   x_Napoli    0.0
8                   x_Rostov    0.0
9                x_Rotterdam    0.0
10            x_StPetersburg    1.0
11           y_Berlin_Moscow    0.0
12           y_Berlin_Rostov    0.0
13     y_Berlin_StPetersburg    0.0
14           y_Boston_Berlin    0.0
15          y_Boston_Hamburg    0.0
16         y_Boston_Istanbul    0.0
17           y_Boston_London    1.0
18           y_Boston_Napoli    0.0
19        y_Boston_Rotterdam    0.0
20          y_Hamburg_Moscow    0.0
21          y_Hamburg_Rostov    0.0
22    y_Hamburg_StPetersburg    0.0
23         y_Istanbul_Moscow    0.0
24         y_Istanbul_Rostov    0.0
25   y_Istanbul_StPetersburg    0.0
26     y_Jacksonville_Berlin    0.0
27    y_Jacksonville_Hamburg    0.0
28   y_Jacksonville_Istanbul    0.0
29     y_Jacksonville_London    0.0
30     y_Jacksonville_Napoli    0.0
31  y_Jacksonville_Rotterdam    0.0
32           y_London_Moscow    0.0
33           y_London_Rostov    0.0
34     y_London_StPetersburg    1.0
35           y_Napoli_Moscow    0.0
36           y_Napoli_Rostov    0.0
37     y_Napoli_StPetersburg    0.0
38        y_Rotterdam_Moscow    0.0
39        y_Rotterdam_Rostov    0.0
40  y_Rotterdam_StPetersburg    0.0

The above result confirms what we already knew from section 1.1 - the fastest path from Boston to StPetersburg is **Boston --> London --> Saint Petersburg (12.70 hr)**. We can solve similar fomulations for each source-target pair to get the desired results.

<!-- It is worth noting that most BIP solvers use some variation of the Branch and Bound algorithm which recursively solves repeated LP relaxations of the original BIP with the simplex method. Since the simplex algorithm is a polynomial time algorithm, the average total running time for the original BIP is close to polynomial in the size of the problem (i.e. number of variables). Let $N: = $number of vertices and $M:=$ number of edges in the graph. Then, from the above formulation, we note that the total number of variables is $M + N$. The average running time of the Branch and Bound algorithm is given by $O(M + N)^p$, $p > 1$ 

On the other hand, Dijkstra's shortest path algorithm has a running time of $O(MN)$. -->

## 1.3 Minimum cost flow approach

## Seervada Park Problem

In [28]:
import cvxpy as cp

## Seervada problem
# parameters
M = 1000000
xO = 1
xT = 1
cOA = 2
cOB = 5
cOC = 4
cAB = 2
cAD = 7
cBC = 1
cBE = 3
cBD = 4
cCE = 4
cED = 1
cDT = 5
cET = 7

# variables
xA = cp.Variable(1, boolean=True, name='xA')
xB = cp.Variable(1, boolean=True, name='xB')
xC = cp.Variable(1, boolean=True, name='xC')
xD = cp.Variable(1, boolean=True, name='xD')
xE = cp.Variable(1, boolean=True, name='xE')

yOA = cp.Variable(1, boolean=True, name='yOA')
yOB = cp.Variable(1, boolean=True, name='yOB')
yOC = cp.Variable(1, boolean=True, name='yOC')
yAB = cp.Variable(1, boolean=True, name='yAB')
yAD = cp.Variable(1, boolean=True, name='yAD')
yBC = cp.Variable(1, boolean=True, name='yBC')
yBE = cp.Variable(1, boolean=True, name='yBE')
yBD = cp.Variable(1, boolean=True, name='yBD')
yCE = cp.Variable(1, boolean=True, name='yCE')
yDT = cp.Variable(1, boolean=True, name='yDT')
yED = cp.Variable(1, boolean=True, name='yED')
yET = cp.Variable(1, boolean=True, name='yET')

# constraints
outbound_cons = [
    yOA + yOB + yOC <= 1 + M * (1 - xO),
    yOA + yOB + yOC >= 1 - M * (1 - xO),

    yAB + yAD <= 1 + M * (1 - xA),
    yAB + yAD >= 1 - M * (1 - xA),

    yBC + yBD + yBE <= 1 + M * (1 - xB),
    yBC + yBD + yBE >= 1 - M * (1 - xB),

    yCE <= 1 + M * (1 - xC),
    yCE >= 1 - M * (1 - xC),

    yDT <= 1 + M * (1 - xD),
    yDT >= 1 - M * (1 - xD),

    yED + yET <= 1 + M * (1 - xE),
    yED + yET >= 1 - M * (1 - xE),
]

inbound_cons = [
    yOA <= 1 + M * (1 - xA),
    yOA >= 1 - M * (1 - xA),
    
    yAB + yOB <= 1 + M * (1 - xB),
    yAB + yOB >= 1 - M * (1 - xB),
    
    yOC + yBC <= 1 + M * (1 - xC),
    yOC + yBC >= 1 - M * (1 - xC),

    yAD + yBD + yED <= 1 + M * (1 - xD),
    yAD + yBD + yED >= 1 - M * (1 - xD),

    yBE + yCE <= 1 + M * (1 - xE),
    yBE + yCE >= 1 - M * (1 - xE),

    yDT + yET <= 1 + M * (1 - xT),
    yDT + yET >= 1 - M * (1 - xT),
]

connectivity_cons = [
    xO + xA >= 2 * yOA,
    xO + xB >= 2 * yOB,
    xO + xC >= 2 * yOC,
    xA + xB >= 2 * yAB,
    xA + xD >= 2 * yAD,
    xB + xC >= 2 * yBC,
    xB + xE >= 2 * yBE,
    xB + xD >= 2 * yBD,
    xC + xE >= 2 * yCE,
    xE + xD >= 2 * yED,
    xD + xT >= 2 * yDT,
    xE + xT >= 2 * yET,
]

constraints = outbound_cons + inbound_cons + connectivity_cons
# objective

obj = cp.Minimize(
    cOA * yOA + 
    cOB * yOB + 
    cOC * yOC + 
    cAB * yAB + 
    cAD * yAD + 
    cBC * yBC + 
    cBE * yBE + 
    cBD * yBD +
    cCE * yCE +
    cDE * yED +
    cDT * yDT +
    cET * yET
)

In [29]:
prob = cp.Problem(obj, constraints)

In [30]:
prob.solve()

13.0

In [31]:
from solver.utils import get_result_summary

In [32]:
get_result_summary(prob)

{'status': 'optimal',
 'optimal_value': 13.0,
 'optimal_solution': {'yOA': 1.0,
  'yOB': 0.0,
  'yOC': 0.0,
  'yAB': 1.0,
  'yAD': 0.0,
  'yBC': 0.0,
  'yBE': 0.0,
  'yBD': 1.0,
  'yCE': 0.0,
  'yED': 0.0,
  'yDT': 1.0,
  'yET': 0.0,
  'xA': 1.0,
  'xB': 1.0,
  'xC': 0.0,
  'xD': 1.0,
  'xE': 0.0}}

# The maximum flow problem
## *As a linear program*

<img src="./images/seervada_maxflow.png" width="500" length="500" align="center">

In [9]:
# variables
xOA = cp.Variable(1, nonneg = True, name='xOA')
xOB = cp.Variable(1, nonneg = True, name='xOB')
xOC = cp.Variable(1, nonneg = True, name='xOC')

xAB = cp.Variable(1, nonneg = True, name='xAB')
xAD = cp.Variable(1, nonneg = True, name='xAD')

xBC = cp.Variable(1, nonneg = True, name='xBC')
xBD = cp.Variable(1, nonneg = True, name='xBD')
xBE = cp.Variable(1, nonneg = True, name='xBE')

xCE = cp.Variable(1, nonneg = True, name='xCE')

xDT = cp.Variable(1, nonneg = True, name='xDT')

xED = cp.Variable(1, nonneg = True, name='xED')
xET = cp.Variable(1, nonneg = True, name='xET')

In [11]:
# constraints
node_flow_cons = [
    xOA - xAB - xAD == 0,
    xOB + xAB - xBC - xBE - xBD == 0,
    xOC + xBC - xCE == 0,
    xAD + xBD + xED - xDT == 0,
    xCE + xBE - xED - xET == 0,
]

total_flow_cons = [-xOA - xOB - xOC + xDT + xET == 0,]

edge_capacity_cons = [
    xOA <= 5,
    xOB <= 7,
    xOC <= 4,
    xAB <= 1,
    xAD <= 3,
    xBD <= 4,
    xBE <= 5,
    xBC <= 2,
    xCE <= 4,
    xDT <= 9,
    xET <= 6,
    xED <= 1,
]

In [12]:
obj = cp.Maximize(xDT + xET)

In [13]:
cons = node_flow_cons + total_flow_cons + edge_capacity_cons

In [14]:
prob = cp.Problem(obj, cons)

In [15]:
prob.solve()

13.999999996621327